File: tools/delta_report.py
#!/usr/bin/env python3
"""Integration delta report.

Usage:
  python3 tools/delta_report.py mappings/iso42001-2023--iso27001-2022.yaml
  python3 tools/delta_report.py mappings/eu-ai-act-2024-2026-1744--iso42001-2023.yaml

For the mapping set's from_framework, reports each requirement's best
coverage by mapping edges (to_framework may span inherited namespaces):
  covered   - has a satisfied_by edge
  partial   - best edge is partially_satisfied_by (gap notes shown)
  adjacent  - only related/supports edges
  net-new   - no edges at all

Used for tier residuals (ISMS→AIMS, AIMS→AI Act); the product's roadmap
generator consumes the same logic.
"""

import glob
import os
import sys

import yaml

RANK = {"satisfied_by": 3, "partially_satisfied_by": 2, "supports": 1, "related": 1}
LABEL = {3: "covered", 2: "partial", 1: "adjacent", 0: "net-new"}


def main(map_path):
    root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    ms = yaml.safe_load(open(map_path))
    from_fw = ms["from_framework"]  # e.g. iso42001/2023

    fw, ver = from_fw.split("/")
    reqs = {}
    for p in sorted(glob.glob(os.path.join(root, "frameworks", fw, ver, "requirements", "*.yaml"))):
        d = yaml.safe_load(open(p))
        reqs[d["id"]] = d["title"]

    best = {rid: 0 for rid in reqs}
    notes = {rid: [] for rid in reqs}
    for m in ms["mappings"]:
        rid = m["from"]
        if rid not in best:
            continue
        r = RANK.get(m["relation"], 0)
        best[rid] = max(best[rid], r)
        if m["relation"] == "partially_satisfied_by" and m.get("note"):
            notes[rid].append(m["note"].strip().replace("\n", " "))

    buckets = {"covered": [], "partial": [], "adjacent": [], "net-new": []}
    for rid, title in reqs.items():
        buckets[LABEL[best[rid]]].append((rid, title))

    print(f"Integration delta: {from_fw} against {ms['to_framework']}")
    print(f"{len(reqs)} requirements: "
          + ", ".join(f"{k}={len(v)}" for k, v in buckets.items()) + "\n")
    for bucket in ["net-new", "partial", "adjacent", "covered"]:
        if not buckets[bucket]:
            continue
        print(f"== {bucket} ({len(buckets[bucket])}) ==")
        for rid, title in buckets[bucket]:
            print(f"  {rid.split('/')[-1]:>8}  {title}")
            if bucket == "partial":
                for n in notes[rid]:
                    gap = n.split("Gap:")[-1].strip().rstrip(".") if "Gap:" in n else n
                    print(f"            gap: {gap}")
        print()


if __name__ == "__main__":
    try:
        sys.exit(main(sys.argv[1]) or 0)
    except BrokenPipeError:
        os._exit(0)  # graceful under `| head`

File: lib/isomer/delta_report.ex
defmodule Isomer.DeltaReport do
  @moduledoc "Integration delta report (port of tools/delta_report.py)."

  alias Isomer.Paths
  alias Isomer.YAML

  @rank %{
    "satisfied_by" => 3,
    "partially_satisfied_by" => 2,
    "supports" => 1,
    "related" => 1
  }

  @label %{3 => :covered, 2 => :partial, 1 => :adjacent, 0 => :net_new}

  def run(map_path, root \\ Isomer.root()) do
    root = Paths.expand_root!(root)
    map_path = Paths.expand_under!(root, map_path)
    ms = YAML.read!(map_path)
    from_fw = ms["from_framework"]
    [fw, ver] = String.split(from_fw, "/", parts: 2)
    fw = Paths.segment!(fw)
    ver = Paths.segment!(ver)

    reqs =
      root
      |> Paths.wildcard!("frameworks/#{fw}/#{ver}/requirements/*.yaml")
      |> Map.new(fn p ->
        d = YAML.read!(p)
        {d["id"], d["title"]}
      end)

    best = Map.new(reqs, fn {rid, _} -> {rid, 0} end)
    notes = Map.new(reqs, fn {rid, _} -> {rid, []} end)

    {best, notes} =
      Enum.reduce(ms["mappings"] || [], {best, notes}, fn m, {b, n} ->
        rid = m["from"]

        if Map.has_key?(b, rid) do
          r = Map.get(@rank, m["relation"], 0)
          b = Map.update!(b, rid, &max(&1, r))

          n =
            if m["relation"] == "partially_satisfied_by" and is_binary(m["note"]) do
              note =
                m["note"]
                |> String.trim()
                |> String.replace("\n", " ")

              Map.update!(n, rid, &[note | &1])
            else
              n
            end

          {b, n}
        else
          {b, n}
        end
      end)

    buckets = %{covered: [], partial: [], adjacent: [], net_new: []}

    buckets =
      Enum.reduce(reqs, buckets, fn {rid, title}, acc ->
        label = Map.fetch!(@label, Map.fetch!(best, rid))
        Map.update!(acc, label, &[{rid, title} | &1])
      end)

    buckets = Map.new(buckets, fn {k, v} -> {k, Enum.reverse(v)} end)

    IO.puts("Integration delta: #{from_fw} against #{ms["to_framework"]}")

    IO.puts(
      "#{map_size(reqs)} requirements: " <>
        Enum.map_join([:covered, :partial, :adjacent, :net_new], ", ", fn k ->
          "#{k}=#{length(buckets[k])}"
        end)
    )

    IO.puts("")

    for bucket <- [:net_new, :partial, :adjacent, :covered],
        rows = buckets[bucket],
        rows != [] do
      IO.puts("== #{bucket} (#{length(rows)}) ==")

      for {rid, title} <- rows do
        ref = rid |> String.split("/") |> List.last()
        IO.puts("  #{String.pad_leading(ref, 8)}  #{title}")

        if bucket == :partial do
          for n <- Enum.reverse(notes[rid]) do
            gap =
              if String.contains?(n, "Gap:") do
                n
                |> String.split("Gap:")
                |> List.last()
                |> String.trim()
                |> String.trim_trailing(".")
              else
                n
              end

            IO.puts("            gap: #{gap}")
          end
        end
      end

      IO.puts("")
    end

    :ok
  end
end
